tt_metal/impl/buffers/global_circular_buffer.hpp

```diff
@@ -26,13 +26,18 @@ namespace v1 {
 namespace experimental {
 
 class GlobalCircularBuffer {
+    struct Private {
+        explicit Private() = default;
+    };
+
 public:
     GlobalCircularBuffer(
         Device* device,
         const std::unordered_map<CoreCoord, CoreRangeSet>& sender_receiver_core_mapping,
         uint32_t size,
         BufferType buffer_type,
-        tt::stl::Span<const SubDeviceId> sub_device_ids);
+        tt::stl::Span<const SubDeviceId> sub_device_ids,
+        Private);
```

Comment:
Make the ctor itself private instead?

Same for `GlobalSemaphore`

Reply:
This was following the pattern that was added for the Buffer class, which enables using std::make_shared<T> instead of std::shared_ptr<T>(new T) in the creation functions.

Reply:
@sminakov-tt thoughts on this as you originally added this pattern for buffer?

Reply:
By thoughts are basically the same as you described. It's the suggested way to be used for enable_shared_from_this: https://en.cppreference.com/w/cpp/memory/enable_shared_from_this.

But can you please move this "private" constructor closer to the bottom of the class and add a comment.

Reply:
Done. I've moved the creation fns to the top, constructors to the bottom and a comment saying they're "private" and to use the create fns instead.

Reply:
Thanks, I haven't noticed the `shared_ptr`! Why are we switching to the shared ownership though? If this is really desired / needed, another option is to create a `shared_ptr` using the move constructor at the client side, so that we are not imposing this on the user.

Reply:
^ s/switching/using - I see this is the existing usage.